### ISSUE TITLE
Rest API returning 404 when trying to get product's documents

### ISSUE THREAD

**Issue by tm-newslang** [Dolibarr/dolibarr#20558]
### Bug

When trying to access product documents via the API with the following URL, either using the ID :

/api/index.php/documents?modulepart=product&id=13242

Or the product's reference :

/api/index.php/documents?modulepart=product&ref=1252522

It finds the product alright, but returns a 404 error saying it doesn't return any document, when they are several documents for this product :

```
{
  "error": {
    "code": 404,
    "message": "Not Found: Search for modulepart product with Id 13242 or Ref 1252522 does not return any document."
  },
  "debug": {
    "source": "api_documents.class.php:467 at call stage",
    "stages": {
      "success": [
        "get",
        "route",
        "negotiate",
        "authenticate",
        "validate"
      ],
      "failure": [
        "call",
        "message"
      ]
    }
  }
}
```

It's happening in version 14.0.5 in my case, but was also happening in version 13.0.0 and 13.0.1 according to this thread : 
https://www.dolibarr.fr/forum/t/api-rest-404-photo-produit/35562/

I didn't see anything about this in the changelog so I'm assuming this has not been addressed. I'm testing feature for a potential development for a client and I cannot update Dolibarr to check this out for now.

If there's another way to return documents from a product (in my case : product's images) I didn't find it in the doc, but if please correct me if my method is wrong.

### Environment Version

14.0.5

### Environment OS

Linux h2web18 4.9.0-0.bpo.12-amd64 #1 SMP Debian 4.9.210-1~deb8u1 (2020-02-21) x86_64

### Environment Web server

Apache

### Environment PHP

7.4.28

### Environment Database

MySQL or MariaDB 5.6.50-log

### Expected and actual behavior

It should return a list of documents linked to this product.

### Steps to reproduce the behavior

Activate the REST API module in Dolibarr, create a unique token for your login, and then simply execute this API call in any browser :

your-dolibarr-url/api/index.php/documents?modulepart=product&id=[id of a product with documents]

This should return a 404 even though documents exist for this product.

### Attached files

<img width="1154" alt="image" src="https://user-images.githubusercontent.com/59600599/161902958-69043523-24b7-469f-8d2b-693f22ae20f7.png">

The aforementioned product list of documents.

**Comment by tm-newslang** [comment]
I'm following up on this problem, after a few exchanges on the subject in the Dolibarr forum it seems to be related to the way Dolibarr tries to get the folder where the documents images are stored. In the case of my client (working now with version 15.0.1), it's not looking in the right folder.

I found out in this file :

https://github.com/Dolibarr/dolibarr/blob/develop/htdocs/api/class/api_documents.class.php

On line 417, the following code :

```
...
} elseif ($modulepart == 'produit' || $modulepart == 'product') {
			require_once DOL_DOCUMENT_ROOT.'/product/class/product.class.php';

			if (!DolibarrApiAccess::$user->rights->produit->lire) {
				throw new RestException(401);
			}

			$object = new Product($this->db);
			$result = $object->fetch($id, $ref);
			if ($result == 0) {
				throw new RestException(404, 'Product not found');
			} elseif ($result < 0) {
				throw new RestException(500, 'Error while fetching object: '.$object->error);
			}

			$upload_dir = $conf->product->multidir_output[$object->entity].'/'.get_exdir(0, 0, 0, 1, $object, 'product');
		} elseif ($modulepart == 'agenda' || $modulepart == 'action' || $modulepart == 'event') {
...
```

This code is supposed to get the upload dir where the documents should be saved for the current product. I took a look at the returned value, it's the following :

**…/dolibarr/documents/produit/1252522**

Except, in my client's configuration, this folder doesn't exist. Inside the folder **…/dolibarr/documents/produit** I have a whole bunche of folders, from 0 to 9, and in each of them the same thing : numbered folders.

It seems like the folder or organized to reflect the last two values ofr the ids of the product. In my case, this product's documents (id 13242) are inside the following folder :

**…/dolibarr/documents/produit/2/4/13242/photos**

Where 2/4 are simply 42 in reverse.

To summarize, `$upload_dir` doesn't contain the right path. The following line of code :

`$conf->product->multidir_output[$object->entity]`

Returns **/dolibarr/documents/produit** which is correct, but the method `get_exdir` which should get the rest of the path, specific to the given product, is not working properly. 

I'm not saying it's always the case, maybe it's a set back from updating from older versions. But even with brand new products (I tried creating one) the API returns a 404 and the directory returned is incorrect.

Is this of any help ? Can somebody help me fix this problem ?

**Comment by fappels** [comment]
You better run 'scripts/product/migrate_picture_path.php'. Seems API does not support global constant PRODUCT_USE_OLD_PATH_FOR_PHOTO = 1. Path system changed 7 years ago and probably support will be removed in future.

**Comment by tm-newslang** [comment]
Thanks for the feedback, this is a client env and they only have dolibarr htdocs installed. How can I go about running this command in their env ? Or can I simply copy this file :

https://doxygen.dolibarr.org/develop/d8/d25/product_2migrate__picture__path_8php_source.html

And paste it in the dolibarr folder and run it : `php migrate_picture_path.php` ?

**Comment by tm-newslang** [comment]
I copied the source of migrate_picture_path.php and starting changing it first take a look at what was happening, it transformed photos path from this, for product id **10731** with ref **180.1290.0510** :

`.../dolibarr/documents/produit/1/3/10731/photos/`

To this :

`.../dolibarr/documents/produit/180.1290.0510`

And after running the script as a test on this specific product, I saw that images where not visible anymore for this product. I had to go in the Dolibarr config change PRODUCT_USE_OLD_PATH_FOR_PHOTO to 0 to make it reappear for this product. Of course, it then didn't work anymore for the other product.

As a precaution I slightly changed the script from "dol_move" to "dol_copy" in order to keep a copy of the original photos in their right place. But otherwise, after running this, it worked and the API was able to get the documents.

Thanks @fappels for the help, it seems that the API isn't working with PRODUCT_USE_OLD_PATH_FOR_PHOTO = 1 as you mentioned.

**Comment by awsv-dev** [comment]
Hi,

I apologize for revisiting this discussion, but I wanted to raise this topic again since the issue with retrieving product images doesn’t seem to be fully resolved.

I’m currently facing the same problem while developing an app, and after reviewing the code from @tm-newslang and comparing it with my Dolibarr v21.0.1 installation, I noticed an inconsistency in the generated folder path ([forum](https://www.dolibarr.fr/forum/t/api-rest-404-photo-produit/35562/21)).

Patch from @tm-newslang:
`
// start patch
$id = $object->id;
$upload_dir = $conf->product->multidir_output[$object->entity].'/'.substr($id, -1).'/'.substr($id, -2, 1).'/'.$id.'/photos';
// end patch
`

Currently, the code returns this path:
`"htdocs/documents/produit/3/3/3/photos"`

As you can see, the folder structure is /3/3/3, when it should actually be /3/0/3. The correct structure should include 0 as the default folder for files, followed by the product ID ($id), and finally the /photos string, as shown in this example:

`$upload_dir = $conf->product->multidir_output[$object->entity].'/'.get_exdir(0, 0, 0, 1, $object, 'product') . '/' . $id . '/photos';`

And this returns the path `"htdocs/documents/produit/3/0/3/photos"`

From the [api_documents.class.php](https://github.com/Dolibarr/dolibarr/blob/develop/htdocs/api/class/api_documents.class.php)

![Image](https://github.com/user-attachments/assets/45630f81-5b9c-403c-b9af-db2085a382f0)

To fix this, you simply need to append `. '/' . $id . '/photos` at the end of the relevant line (obviously before the semicolon 😊).

et c'est

From Swagger UI
![Image](https://github.com/user-attachments/assets/c6487b23-aaed-4efa-8ad8-a2fb88f4907f)

From HTTPie
![Image](https://github.com/user-attachments/assets/7fc3b2e9-72e8-4c3e-b5d9-3b50421b5c67)
